packages/lesswrong/lib/cookies/utils.ts

```diff
@@ -0,0 +1,94 @@
+import { getExplicitConsentRequiredAsync } from "../../components/common/CookieBanner/geolocation";
+
+export const CookiesTable: Record<string, CookieSignature> = {};
+
+const CookieTypes = new TupleSet(["necessary", "functional", "analytics"] as const)
+export type CookieType = UnionOf<typeof CookieTypes>;
+
+export const COOKIE_PREFERENCES_COOKIE = registerCookie({
+  name: "cookie_preferences", // Must match variable in Google Tag Manager
+  type: "necessary",
+  description: "Stores the current cookie preferences (set by the user, or automatically if outside a country subject to GDPR)",
+});
+
+export const COOKIE_CONSENT_TIMESTAMP_COOKIE = registerCookie({
+  name: "cookie_consent_timestamp",
+  type: "necessary",
+  description: "Stores the time at which the user set their cookie preferences (once this is set the cookie preferences will never be changed automatically)",
+});
+
+export type CookieSignature = {
+  name: string;
+  type: CookieType;
+  /** User readable description of what the cookie does (shown in cookie policy) */
+  description: string;
+  thirdPartyName?: string;
+  /** String description of the longest possible expiry date. Can be e.g. "12 months" or "Session" */
+  maxExpires: string;
+  matches: (name: string) => boolean;
+}
+// maxExpires and matches are filled in by registerCookie if not provided
+export type CookieSignatureMinimum = Omit<CookieSignature, 'maxExpires' | 'matches'> & {
+  matches?: (name: string) => boolean;
+  maxExpires?: string;
+}
+
+export const isValidCookieTypeArray = (types?: string[] | null): types is CookieType[] => {
+  if (!types) return false;
+  return types.every(type => CookieTypes.has(type));
+}
+
+export function isCookieAllowed(name: string, cookieTypesAllowed: CookieType[]): boolean {
+  const cookie = CookiesTable[name];
+  if (cookie) {
+    return cookieTypesAllowed.includes(cookie.type);
+  }
+
+  // call the `matches` function of all cookies to see if it matches
+  for (const cookieName in CookiesTable) {
+    if (CookiesTable[cookieName].matches(name)) {
+      return cookieTypesAllowed.includes(CookiesTable[cookieName].type);
+    }
+  }
+  throw new Error(`Unknown cookie: ${name}, you must register it with registerCookie`);
+}
+
+/**
+ * Non-hook version of useCookiePreferences for use outside of React. Returns the list of
+ * allowed cookie types and whether the user has given explicit consent (if they are outside
+ * the EU, this will always be true)
+ */
+export async function getCookiePreferences(): Promise<{
+  cookiePreferences: CookieType[];
+  explicitConsentGiven: boolean;
+}> {
+  const cookies = new Cookies();
+  const preferencesValue = cookies.get(COOKIE_PREFERENCES_COOKIE);
+  const consentTimestamp = cookies.get(COOKIE_CONSENT_TIMESTAMP_COOKIE);
+  const explicitConsentGiven = !!consentTimestamp && isValidCookieTypeArray(preferencesValue);
+
+  const explicitConsentRequired = await getExplicitConsentRequiredAsync();
```

Comment:
I would rather the sync one didn't exist at all, but it's necessary in the case where it's used in a react component to prevent extra renders (you have to render at least once before getting the result of an async call). This function is only called from within another async function so it's simpler to just use the async one (which still checks the cache)

Reply:
Oh I missed that it checks the cache too, seems reasonable